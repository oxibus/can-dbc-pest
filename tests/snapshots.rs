#![cfg(feature = "encoding")]

use std::borrow::Cow;
use std::fs;
use std::path::Path;

use can_dbc_pest::{decode_cp1252, DbcParser, Rule};
use insta::{assert_debug_snapshot, with_settings};
use pest::Parser;

/// Test parsing all DBC files
#[test]
fn test_cantools_dbc_files() {
    let test_dirs = [
        (
            "tests/shared-test-files/dbc-cantools",
            "dbc-cantools",
            true,
            true,
        ),
        ("tests/shared-test-files/canpy", "canpy", true, true),
        ("tests/opendbc/opendbc/dbc", "opendbc", false, false),
    ];
    for (path, file_name, snapshot, cp1251) in test_dirs {
        let path = Path::new(path);
        let snapshot_path = String::from("snapshots-") + file_name;
        // snapshots should go to /tests/snapshots-* directory
        with_settings! {
            { omit_expression => true,
              snapshot_path => snapshot_path ,
              prepend_module_to_snapshot => false },
            {
                test_dbc_files(path, snapshot, cp1251);
            }
        }
    }
}

/// Test parsing all DBC files in the given directory.
fn test_dbc_files(dir: impl AsRef<Path>, snapshot: bool, cp1251: bool) {
    let dir = dir.as_ref();
    let dir_display = dir.display();
    let dir_content = fs::read_dir(dir).unwrap_or_else(|e| {
        panic!(
            "
--------------------------------------------------------------------------
Error reading dbc test files from   {dir_display}
{e}
Make sure git submodules are up to date by running
    git submodule update --init --recursive
--------------------------------------------------------------------------
"
        )
    });

    eprintln!("Testing dbc files in directory: {dir_display}");
    for dbc_path in dir_content {
        let path = dbc_path.unwrap().path();
        if path
            .extension()
            .is_some_and(|ext| ext.eq_ignore_ascii_case("dbc"))
        {
            parse_one_file(path, snapshot, cp1251);
        }
    }
}

/// Parse a single DBC file and assert a snapshot of the result.
fn parse_one_file(path: impl AsRef<Path>, snapshot: bool, cp1251: bool) {
    let path = path.as_ref();
    eprintln!("Testing DBC file: {}", path.display());
    let file_name = path.file_stem().unwrap().to_string_lossy().to_string();
    let buffer = fs::read(path).unwrap();
    let buffer = if cp1251 {
        decode_cp1252(&buffer)
            .unwrap_or_else(|| panic!("Failed to decode {} as cp1252", path.display()))
    } else {
        Cow::Borrowed(
            std::str::from_utf8(&buffer)
                .unwrap_or_else(|_| panic!("Failed to decode {} as utf-8", path.display())),
        )
    };

    match DbcParser::parse(Rule::file, &buffer) {
        Ok(pairs) => {
            if std::env::var("SKIP_INSTA").is_err()
                && (snapshot || std::env::var("FORCE_INSTA").is_ok())
            {
                assert_debug_snapshot!(file_name, pairs);
            }
        }
        Err(e) => panic!("Failed to parse {file_name}.dbc: {e:#?}"),
    }
}
