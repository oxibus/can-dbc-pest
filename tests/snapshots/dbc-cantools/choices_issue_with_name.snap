---
source: tests/snapshots.rs
---
[
    Pair {
        rule: file,
        span: Span {
            str: "VERSION \"\"\n\n\nNS_ : \n\tNS_DESC_\n\tCM_\n\tBA_DEF_\n\tBA_\n\tVAL_\n\tCAT_DEF_\n\tCAT_\n\tFILTER\n\tBA_DEF_DEF_\n\tEV_DATA_\n\tENVVAR_DATA_\n\tSGTYPE_\n\tSGTYPE_VAL_\n\tBA_DEF_SGTYPE_\n\tBA_SGTYPE_\n\tSIG_TYPE_REF_\n\tVAL_TABLE_\n\tSIG_GROUP_\n\tSIG_VALTYPE_\n\tSIGTYPE_VALTYPE_\n\tBO_TX_BU_\n\tBA_DEF_REL_\n\tBA_REL_\n\tBA_DEF_DEF_REL_\n\tBU_SG_REL_\n\tBU_EV_REL_\n\tBU_BO_REL_\n\tSG_MUL_VAL_\n\nBS_:\n\nBU_:\nVAL_TABLE_ CmdResp 1 \"SignalWithChoices_CmdRespOK\" 0 \"SignalWithChoices_CmdRespErr\" ;\n\n\n\nBO_ 2148073472 TestMessage: 1 Vector__XXX\n SG_ SignalWithChoices : 0|1@1+ (1,0) [0|0] \"\" Vector__XXX\n\n\n\nBA_DEF_  \"BusType\" STRING ;\nBA_DEF_DEF_  \"BusType\" \"CAN\";\nVAL_ 2148073472 SignalWithChoices 1 \"SignalWithChoices_CmdRespOK\" 0 \"SignalWithChoices_CmdRespErr\" ;\n",
            start: 0,
            end: 700,
        },
        inner: [
            Pair {
                rule: version,
                span: Span {
                    str: "VERSION \"\"\n",
                    start: 0,
                    end: 11,
                },
                inner: [
                    Pair {
                        rule: quoted_str,
                        span: Span {
                            str: "\"\"",
                            start: 8,
                            end: 10,
                        },
                        inner: [
                            Pair {
                                rule: string,
                                span: Span {
                                    str: "",
                                    start: 9,
                                    end: 9,
                                },
                                inner: [],
                            },
                        ],
                    },
                ],
            },
            Pair {
                rule: new_symbols,
                span: Span {
                    str: "NS_ : \n\tNS_DESC_\n\tCM_\n\tBA_DEF_\n\tBA_\n\tVAL_\n\tCAT_DEF_\n\tCAT_\n\tFILTER\n\tBA_DEF_DEF_\n\tEV_DATA_\n\tENVVAR_DATA_\n\tSGTYPE_\n\tSGTYPE_VAL_\n\tBA_DEF_SGTYPE_\n\tBA_SGTYPE_\n\tSIG_TYPE_REF_\n\tVAL_TABLE_\n\tSIG_GROUP_\n\tSIG_VALTYPE_\n\tSIGTYPE_VALTYPE_\n\tBO_TX_BU_\n\tBA_DEF_REL_\n\tBA_REL_\n\tBA_DEF_DEF_REL_\n\tBU_SG_REL_\n\tBU_EV_REL_\n\tBU_BO_REL_\n\tSG_MUL_VAL_\n\n",
                    start: 13,
                    end: 337,
                },
                inner: [
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "NS_DESC_",
                            start: 21,
                            end: 29,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "CM_",
                            start: 31,
                            end: 34,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "BA_DEF_",
                            start: 36,
                            end: 43,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "BA_",
                            start: 45,
                            end: 48,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "VAL_",
                            start: 50,
                            end: 54,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "CAT_DEF_",
                            start: 56,
                            end: 64,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "CAT_",
                            start: 66,
                            end: 70,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "FILTER",
                            start: 72,
                            end: 78,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "BA_DEF_DEF_",
                            start: 80,
                            end: 91,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "EV_DATA_",
                            start: 93,
                            end: 101,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "ENVVAR_DATA_",
                            start: 103,
                            end: 115,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "SGTYPE_",
                            start: 117,
                            end: 124,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "SGTYPE_VAL_",
                            start: 126,
                            end: 137,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "BA_DEF_SGTYPE_",
                            start: 139,
                            end: 153,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "BA_SGTYPE_",
                            start: 155,
                            end: 165,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "SIG_TYPE_REF_",
                            start: 167,
                            end: 180,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "VAL_TABLE_",
                            start: 182,
                            end: 192,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "SIG_GROUP_",
                            start: 194,
                            end: 204,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "SIG_VALTYPE_",
                            start: 206,
                            end: 218,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "SIGTYPE_VALTYPE_",
                            start: 220,
                            end: 236,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "BO_TX_BU_",
                            start: 238,
                            end: 247,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "BA_DEF_REL_",
                            start: 249,
                            end: 260,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "BA_REL_",
                            start: 262,
                            end: 269,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "BA_DEF_DEF_REL_",
                            start: 271,
                            end: 286,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "BU_SG_REL_",
                            start: 288,
                            end: 298,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "BU_EV_REL_",
                            start: 300,
                            end: 310,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "BU_BO_REL_",
                            start: 312,
                            end: 322,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "SG_MUL_VAL_",
                            start: 324,
                            end: 335,
                        },
                        inner: [],
                    },
                ],
            },
            Pair {
                rule: bit_timing,
                span: Span {
                    str: "BS_:\n",
                    start: 337,
                    end: 342,
                },
                inner: [],
            },
            Pair {
                rule: nodes,
                span: Span {
                    str: "BU_:\n",
                    start: 343,
                    end: 348,
                },
                inner: [],
            },
            Pair {
                rule: value_table,
                span: Span {
                    str: "VAL_TABLE_ CmdResp 1 \"SignalWithChoices_CmdRespOK\" 0 \"SignalWithChoices_CmdRespErr\" ;",
                    start: 348,
                    end: 433,
                },
                inner: [
                    Pair {
                        rule: table_name,
                        span: Span {
                            str: "CmdResp",
                            start: 359,
                            end: 366,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: table_value_description,
                        span: Span {
                            str: "1 \"SignalWithChoices_CmdRespOK\"",
                            start: 367,
                            end: 398,
                        },
                        inner: [
                            Pair {
                                rule: table_value,
                                span: Span {
                                    str: "1",
                                    start: 367,
                                    end: 368,
                                },
                                inner: [],
                            },
                            Pair {
                                rule: table_desc,
                                span: Span {
                                    str: "\"SignalWithChoices_CmdRespOK\"",
                                    start: 369,
                                    end: 398,
                                },
                                inner: [
                                    Pair {
                                        rule: string,
                                        span: Span {
                                            str: "SignalWithChoices_CmdRespOK",
                                            start: 370,
                                            end: 397,
                                        },
                                        inner: [],
                                    },
                                ],
                            },
                        ],
                    },
                    Pair {
                        rule: table_value_description,
                        span: Span {
                            str: "0 \"SignalWithChoices_CmdRespErr\"",
                            start: 399,
                            end: 431,
                        },
                        inner: [
                            Pair {
                                rule: table_value,
                                span: Span {
                                    str: "0",
                                    start: 399,
                                    end: 400,
                                },
                                inner: [],
                            },
                            Pair {
                                rule: table_desc,
                                span: Span {
                                    str: "\"SignalWithChoices_CmdRespErr\"",
                                    start: 401,
                                    end: 431,
                                },
                                inner: [
                                    Pair {
                                        rule: string,
                                        span: Span {
                                            str: "SignalWithChoices_CmdRespErr",
                                            start: 402,
                                            end: 430,
                                        },
                                        inner: [],
                                    },
                                ],
                            },
                        ],
                    },
                ],
            },
            Pair {
                rule: message,
                span: Span {
                    str: "BO_ 2148073472 TestMessage: 1 Vector__XXX",
                    start: 437,
                    end: 478,
                },
                inner: [
                    Pair {
                        rule: msg_var,
                        span: Span {
                            str: "BO_ 2148073472",
                            start: 437,
                            end: 451,
                        },
                        inner: [
                            Pair {
                                rule: message_id,
                                span: Span {
                                    str: "2148073472",
                                    start: 441,
                                    end: 451,
                                },
                                inner: [],
                            },
                        ],
                    },
                    Pair {
                        rule: message_name,
                        span: Span {
                            str: "TestMessage",
                            start: 452,
                            end: 463,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: message_size,
                        span: Span {
                            str: "1",
                            start: 465,
                            end: 466,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: transmitter,
                        span: Span {
                            str: "Vector__XXX",
                            start: 467,
                            end: 478,
                        },
                        inner: [],
                    },
                ],
            },
            Pair {
                rule: signal,
                span: Span {
                    str: "SG_ SignalWithChoices : 0|1@1+ (1,0) [0|0] \"\" Vector__XXX",
                    start: 480,
                    end: 537,
                },
                inner: [
                    Pair {
                        rule: signal_name,
                        span: Span {
                            str: "SignalWithChoices",
                            start: 484,
                            end: 501,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: start_bit,
                        span: Span {
                            str: "0",
                            start: 504,
                            end: 505,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: signal_size,
                        span: Span {
                            str: "1",
                            start: 506,
                            end: 507,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: little_endian,
                        span: Span {
                            str: "1",
                            start: 508,
                            end: 509,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: unsigned_type,
                        span: Span {
                            str: "+",
                            start: 509,
                            end: 510,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: factor,
                        span: Span {
                            str: "1",
                            start: 512,
                            end: 513,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: offset,
                        span: Span {
                            str: "0",
                            start: 514,
                            end: 515,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: min_max,
                        span: Span {
                            str: "[0|0]",
                            start: 517,
                            end: 522,
                        },
                        inner: [
                            Pair {
                                rule: minimum,
                                span: Span {
                                    str: "0",
                                    start: 518,
                                    end: 519,
                                },
                                inner: [],
                            },
                            Pair {
                                rule: maximum,
                                span: Span {
                                    str: "0",
                                    start: 520,
                                    end: 521,
                                },
                                inner: [],
                            },
                        ],
                    },
                    Pair {
                        rule: unit,
                        span: Span {
                            str: "\"\"",
                            start: 523,
                            end: 525,
                        },
                        inner: [
                            Pair {
                                rule: string,
                                span: Span {
                                    str: "",
                                    start: 524,
                                    end: 524,
                                },
                                inner: [],
                            },
                        ],
                    },
                    Pair {
                        rule: node_name,
                        span: Span {
                            str: "Vector__XXX",
                            start: 526,
                            end: 537,
                        },
                        inner: [],
                    },
                ],
            },
            Pair {
                rule: attr_def,
                span: Span {
                    str: "BA_DEF_  \"BusType\" STRING ;",
                    start: 541,
                    end: 568,
                },
                inner: [
                    Pair {
                        rule: attribute_name,
                        span: Span {
                            str: "\"BusType\"",
                            start: 550,
                            end: 559,
                        },
                        inner: [
                            Pair {
                                rule: string,
                                span: Span {
                                    str: "BusType",
                                    start: 551,
                                    end: 558,
                                },
                                inner: [],
                            },
                        ],
                    },
                    Pair {
                        rule: attribute_type_string,
                        span: Span {
                            str: "STRING",
                            start: 560,
                            end: 566,
                        },
                        inner: [],
                    },
                ],
            },
            Pair {
                rule: ba_def_def,
                span: Span {
                    str: "BA_DEF_DEF_  \"BusType\" \"CAN\";",
                    start: 569,
                    end: 598,
                },
                inner: [
                    Pair {
                        rule: attribute_name,
                        span: Span {
                            str: "\"BusType\"",
                            start: 582,
                            end: 591,
                        },
                        inner: [
                            Pair {
                                rule: string,
                                span: Span {
                                    str: "BusType",
                                    start: 583,
                                    end: 590,
                                },
                                inner: [],
                            },
                        ],
                    },
                    Pair {
                        rule: num_str_value,
                        span: Span {
                            str: "\"CAN\"",
                            start: 592,
                            end: 597,
                        },
                        inner: [
                            Pair {
                                rule: quoted_str,
                                span: Span {
                                    str: "\"CAN\"",
                                    start: 592,
                                    end: 597,
                                },
                                inner: [
                                    Pair {
                                        rule: string,
                                        span: Span {
                                            str: "CAN",
                                            start: 593,
                                            end: 596,
                                        },
                                        inner: [],
                                    },
                                ],
                            },
                        ],
                    },
                ],
            },
            Pair {
                rule: value_table_def,
                span: Span {
                    str: "VAL_ 2148073472 SignalWithChoices 1 \"SignalWithChoices_CmdRespOK\" 0 \"SignalWithChoices_CmdRespErr\" ;",
                    start: 599,
                    end: 699,
                },
                inner: [
                    Pair {
                        rule: message_id,
                        span: Span {
                            str: "2148073472",
                            start: 604,
                            end: 614,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: signal_name,
                        span: Span {
                            str: "SignalWithChoices",
                            start: 615,
                            end: 632,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: table_value_description,
                        span: Span {
                            str: "1 \"SignalWithChoices_CmdRespOK\"",
                            start: 633,
                            end: 664,
                        },
                        inner: [
                            Pair {
                                rule: table_value,
                                span: Span {
                                    str: "1",
                                    start: 633,
                                    end: 634,
                                },
                                inner: [],
                            },
                            Pair {
                                rule: table_desc,
                                span: Span {
                                    str: "\"SignalWithChoices_CmdRespOK\"",
                                    start: 635,
                                    end: 664,
                                },
                                inner: [
                                    Pair {
                                        rule: string,
                                        span: Span {
                                            str: "SignalWithChoices_CmdRespOK",
                                            start: 636,
                                            end: 663,
                                        },
                                        inner: [],
                                    },
                                ],
                            },
                        ],
                    },
                    Pair {
                        rule: table_value_description,
                        span: Span {
                            str: "0 \"SignalWithChoices_CmdRespErr\"",
                            start: 665,
                            end: 697,
                        },
                        inner: [
                            Pair {
                                rule: table_value,
                                span: Span {
                                    str: "0",
                                    start: 665,
                                    end: 666,
                                },
                                inner: [],
                            },
                            Pair {
                                rule: table_desc,
                                span: Span {
                                    str: "\"SignalWithChoices_CmdRespErr\"",
                                    start: 667,
                                    end: 697,
                                },
                                inner: [
                                    Pair {
                                        rule: string,
                                        span: Span {
                                            str: "SignalWithChoices_CmdRespErr",
                                            start: 668,
                                            end: 696,
                                        },
                                        inner: [],
                                    },
                                ],
                            },
                        ],
                    },
                ],
            },
            Pair {
                rule: EOI,
                span: Span {
                    str: "",
                    start: 700,
                    end: 700,
                },
                inner: [],
            },
        ],
    },
]
